File: GolangSolutions/String/M791_自定义字符串排序.go
/*
   @author Hyperspace
   @date 2022/11/13
   @file M791_自定义字符串排序.go
*/

package String

import "sort"

func customSortString(order string, s string) (res string) {
	cnt := make([]int, 26)
	for _, c := range s {
		cnt[c-'a']++
	}
	for _, c := range order {
		for cnt[c-'a'] > 0 {
			res += string(c)
			cnt[c-'a']--
		}
	}
	for i := range cnt {
		for cnt[i] > 0 {
			res += string(i + 'a')
			cnt[i]--
		}
	}

	return
}

func customSortString2(order string, s string) string {
	cnt := make([]int, 26)
	for _, c := range s {
		cnt[c-'a']++
	}
	var res []rune
	for _, c := range order {
		for cnt[c-'a'] > 0 {
			res = append(res, c)
			cnt[c-'a']--
		}
	}
	for i := range cnt {
		for cnt[i] > 0 {
			res = append(res, rune(i+'a'))
			cnt[i]--
		}
	}

	return string(res)
}

func customSortString3(order string, s string) string {
	weight := map[rune]int{}
	for i, c := range order {
		weight[c] = i + 1
	}
	t := []rune(s)
	sort.Slice(t, func(i, j int) bool {
		return weight[t[i]] < weight[t[j]]
	})

	return string(t)
}

func customSortString4(order string, s string) string {
	weight := make([]int, 26)
	for i, c := range order {
		weight[c-'a'] = i + 1
	}
	t := []rune(s)
	sort.Slice(t, func(i, j int) bool {
		return weight[t[i]-'a'] < weight[t[j]-'a']
	})

	return string(t)
}

File: CPPSolutions/CPPSolutions/M791_自定义字符串排序.cpp
#include <bits\stdc++.h>

using namespace std;

class Solution {
public:
    string customSortString(string order, string s) {
        int cnt[26];
        memset(cnt, 0, sizeof cnt);
        for (auto& c : s) cnt[c - 'a']++;
        string res;
        for (auto& c : order) while (cnt[c - 'a']-- > 0) res += c;
        for (int i = 0; i < 26; i++) while (cnt[i]-- > 0) res += i + 'a';

        return res;
    }

    // 使用string()直接累加同一字母
    string customSortString2(string order, string s) {
        int cnt[26];
        memset(cnt, 0, sizeof cnt);
        for (auto& c : s) cnt[c - 'a']++;
        string res;
        for (auto& c : order) {
            if (cnt[c - 'a'] > 0) {
                res += string(cnt[c - 'a'], c);
                cnt[c - 'a'] = 0;
            }
        }
        for (int i = 0; i < 26; i++) {
            if (cnt[i] > 0) {
                res += string(cnt[i], i + 'a');
                cnt[i] = 0;
            }
        }

        return res;
    }

    // 换用vector定义
    string customSortString3(string order, string s) {
        vector<int> cnt(26);
        for (auto& c : s) cnt[c - 'a']++;
        string res;
        for (auto& c : order) while (cnt[c - 'a']-- > 0) res += c;
        for (int i = 0; i < 26; i++) while (cnt[i]-- > 0) res += i + 'a';

        return res;
    }

    // 自定义排序，按权重升序排列
    string customSortString4(string order, string s) {
        vector<int> weight(26);
        for (int i = 0; i < order.size(); i++) weight[order[i] - 'a'] = i + 1;
        sort(s.begin(), s.end(), [&](char c0, char c1) {
            return weight[c0 - 'a'] < weight[c1 - 'a'];
        });

        return s;
    }
};